### DISCUSSION TITLE
ui.log - support for coloring lines

### DISCUSSION THREAD

**Discussion by whoenig** [zauberzeug/nicegui#2508] [Ideas / Feature Requests]
It would be nice if individual lines in ui.log could be colored (e.g., yellow for warnings, red for errors). 

On the technical side, ui.log currently uses an html textarea, which does not support different text properties, as far as I know. (See https://github.com/zauberzeug/nicegui/blob/main/nicegui/elements/log.py and https://github.com/zauberzeug/nicegui/blob/main/nicegui/elements/log.js).

**Comment by rodja** [comment]
Indeed, this would be helpful. We would be happy to review any pull requests in this direction.

**Comment by falkoschindler** [comment]
I wonder how this could be done conceptually. Should we accept an optional color per line? Or should we add general support for HTML or maybe ANSI escape codes? And is it possible to keep using the existing implementation or should we switch to a more powerful JavaScript library like xterm.js (see https://github.com/zauberzeug/nicegui/discussions/1846)?

**Reply by anoff** [reply]
Personally I prefer ANSI escape codes for compatibility reasons within the Python ecosystem.
In my current usecase I'm using https://pypi.org/project/colored anyway for styling the logs.

**Comment by whoenig** [comment]
On an API level the push function could get a color per line, or stylesheet definition, at least for my use-case. There might be cases where a more general support would be helpful (e.g., error messages in clang can highlight parts of the line to indicate clearly where a problem is).

On the technical side, I am not sure how to even get started. As stated in the initial feature request, the current implementation uses an textarena, which seems to only allow a single style for the whole text.

**Comment by WolfgangFahl** [comment]
I am simply prepending icons:
```python
 LOG_LEVEL_ICONS = {
        logging.DEBUG: "🐞",
        logging.INFO: "ℹ️",
        logging.WARNING: "⚠️",
        logging.ERROR: "❌",
        logging.CRITICAL: "🔥",
    }
```

**Comment by golgitab** [comment]
is it still being developed? 
it seems like changing from Label to Html solves all the problems

```py
from typing import Any, Optional

from ..element import Element
from .label import Label
from .html import Html


class Log(Element, default_classes='nicegui-log'):

    def __init__(self, max_lines: Optional[int] = None) -> None:
        """Log View

        Create a log view that allows to add new lines without re-transmitting the whole history to the client.

        :param max_lines: maximum number of lines before dropping oldest ones (default: `None`)
        """
        super().__init__()
        self.max_lines = max_lines

    def push(self, line: Any) -> None:
        """Add a new line to the log.

        :param line: the line to add (can contain line breaks)
        """
        for text in str(line).splitlines():
            with self:
                Html(text)
        while self.max_lines is not None and len(self.default_slot.children) > self.max_lines:
            self.remove(0)
```

and then i can do like that
```py
log.push(f'<span style="color: red; font-weight: bold;">{msg}<span>') 
```

**Reply by falkoschindler** [reply]
Well, simply using `ui.html` instead of `ui.label` would cause a breaking change. But we could offer a `html: bool = False` parameter for `push()`.

It is still suboptimal because it requires the user to add quite some boilerplate HTML around the actual log message. And it could break the rendering if `msg` would contain characters like `<`.

Actually, this feature request dates back to before PR #2686 introduced a new `ui.log` implementation. Now it would be much easier to simply add a `color: Optional[str] = None` parameter and apply it to the label.

Or maybe `classes: str = ''` would be even better (and more straightforward to implement). This would allow you to call `log.push('Ooops!', classes='text-red')`. Yes, I think this would be my preferred API. What do you think?

**Reply by evnchn** [reply]
Interesting how `ui.log` has `ui.label` inside in a nutshell. I always thought it was something more complex. 

Didn't know much about `ui.log` because I didn't use it much, and only when you show me #2686 I looked into the source code. 

`log.push('Ooops!', classes='text-red')` seems nice, but I wonder if we need also let the user set style and props. 

Also, this may no longer become relevant after we have #4520 (Xterm.js) in NiceGUI. 

Meanwhile I am also considering supporting ANSI colors via Xterm.js natively or maybe https://pypi.org/project/ansi2html/, what do you think? Especially when the goal of colored text may be for displaying other program's console output, which have colors.

**Reply by golgitab** [reply]
> Well, simply using `ui.html` instead of `ui.label` would cause a breaking change. But we could offer a `html: bool = False` parameter for `push()`.
> 
> It is still suboptimal because it requires the user to add quite some boilerplate HTML around the actual log message. And it could break the rendering if `msg` would contain characters like `<`.
> 
> Actually, this feature request dates back to before PR #2686 introduced a new `ui.log` implementation. Now it would be much easier to simply add a `color: Optional[str] = None` parameter and apply it to the label.
> 
> Or maybe `classes: str = ''` would be even better (and more straightforward to implement). This would allow you to call `log.push('Ooops!', classes='text-red')`. Yes, I think this would be my preferred API. What do you think?

I agree that using HTML is not the safest way for users.
I like your smooth suggestion btw.

I think the best way for me to allow user html: bool option as I will be able to control color of timestamp, message and log level separately. Also it would help me to pass some style to the messages like bold, underscore, etc

**Reply by falkoschindler** [reply]
I see. But note that `push` will split the line at \\n which might be weird in combination with `html=True` if the HTML spans multiple lines. We could say that `html=True` will disable `splitlines`, but it would make the behavior inconsistent and harder to predict.

It's surprisingly complicated to cover all use cases:
- simply coloring individual lines e.g. based on severity,
- supporting colored logs from other modules (would require something like ansi2html), or
- giving full control via `html: bool`.

As an alternative you can always clone or subclass `ui.log` since it's really not much code. So maybe we keep `ui.log` nice and simple, while PR #4520 will introduce a more versatile `ui.xterm`.

**Reply by evnchn** [reply]
I think `html=True` disabling `splitlines` is unacceptable, because it makes the number of lines in the `ui.log` suddenly become unpredictable, and the element may grow in height by a lot. 

I think let's keep `ui.log` simple as a container of `ui.label`s per line. Otherwise we'd be breaking user expectations like #2686 (which I am surprised didn't wait until a major release because it could be considered a breaking change)

I recommend: 

- Coloring lines: `ui.log` with passing `classes` to `push`
  - Upcoming PR, should be a minor change. 
- Colored logs from other modules: either `ui.ansi_display` or via `ui.xterm`
  - A new element using `ansi2html`, or #4520 
- Full control via HTML: They can do it themselves, especially since filtering safe HTML doesn't seem like an easy task, and I don't quite want NiceGUI to have a XSS vulnerability. 
  - They can bear their own risk and implement their own element. 

What do you think?

**Reply by falkoschindler** [reply]
I agree 100% 👍🏻

**Reply by golgitab** [reply]
> I see. But note that `push` will split the line at \n which might be weird in combination with `html=True` if the HTML spans multiple lines. We could say that `html=True` will disable `splitlines`, but it would make the behavior inconsistent and harder to predict.
> 
> It's surprisingly complicated to cover all use cases:
> 
> * simply coloring individual lines e.g. based on severity,
> * supporting colored logs from other modules (would require something like ansi2html), or
> * giving full control via `html: bool`.
> 
> As an alternative you can always clone or subclass `ui.log` since it's really not much code. So maybe we keep `ui.log` nice and simple, while PR #4520 will introduce a more versatile `ui.xterm`.

Fully agree, thanks for discussion

**Reply by evnchn** [reply]
Checkout #4750

Should be bare-minimum to get coloring lines with minimal change

**Reply by falkoschindler** [reply]
With #4750 being released, I'm closing this discussion as resolved.
For related features like ANSI support, feel free to open a separate feature request.
